File: android/core/src/main/java/com/lowerbackstretching/core/Pain.kt
package com.lowerbackstretching.core

/**
 * One self-reported pain rating. Pure value contract — the Room-backed
 * `PainLogEntity` in `:app` implements this so the pairing and delta
 * math here can run on either entities or test doubles without dragging
 * Room into `:core`.
 *
 * `painLevel` is 0..10 (no pain..severe). `bodyLocationTag` is a
 * `BodyZone.bodyPartTag` or null when the user skipped the body
 * diagram. `context` is one of [PRE_SESSION], [POST_SESSION].
 */
interface PainMeasurement {
    val recordedAtEpochMillis: Long
    val painLevel: Int
    val bodyLocationTag: String?
    val context: String
}

object PainContext {
    const val PRE_SESSION = "PRE_SESSION"
    const val POST_SESSION = "POST_SESSION"
}

/** A POST-session log paired with its closest preceding PRE log (if any). */
data class SessionPainPair(val pre: PainMeasurement?, val post: PainMeasurement)

/** Numeric view of a paired session — `delta` is null if no pre log was found. */
data class SessionPainDelta(val pre: Int?, val post: Int, val delta: Int?)

fun sessionPainDelta(pair: SessionPainPair): SessionPainDelta =
    SessionPainDelta(
        pre = pair.pre?.painLevel,
        post = pair.post.painLevel,
        delta = pair.pre?.let { pair.post.painLevel - it.painLevel },
    )

/**
 * Pair each POST_SESSION log with the most recent PRE_SESSION log that
 * landed within [lookbackMillis] before it. POST logs without a PRE
 * inside the window come back with `pre = null`. Order of the input
 * list doesn't matter; the result is sorted newest-POST-first.
 */
fun pairSessionPainLogs(
    logs: List<PainMeasurement>,
    lookbackMillis: Long = 60 * 60 * 1000L,
): List<SessionPainPair> {
    val pres = logs
        .filter { it.context == PainContext.PRE_SESSION }
        .sortedBy { it.recordedAtEpochMillis }
    val posts = logs
        .filter { it.context == PainContext.POST_SESSION }
        .sortedByDescending { it.recordedAtEpochMillis }
    return posts.map { post ->
        val pre = pres.lastOrNull { pre ->
            pre.recordedAtEpochMillis <= post.recordedAtEpochMillis &&
                post.recordedAtEpochMillis - pre.recordedAtEpochMillis <= lookbackMillis
        }
        SessionPainPair(pre = pre, post = post)
    }
}

File: ios/LowerBackStretching/Data/Pain.swift
import Foundation

/// Two-valued enum baked into `PainLog.context` so the same table holds
/// both "before" and "after" ratings. Stored as the raw string so
/// existing rows survive a model rename.
enum PainContext {
    static let preSession = "PRE_SESSION"
    static let postSession = "POST_SESSION"
}

/// Pure contract for a recorded pain rating. Lives in the data layer
/// so domain functions can run against test doubles without dragging
/// SwiftData into pure unit tests.
protocol PainMeasurement {
    var recordedAt: Date { get }
    var painLevel: Int { get }
    var bodyLocationTag: String? { get }
    var context: String { get }
}

/// Pairs a POST_SESSION log with the nearest preceding PRE_SESSION log
/// (if one exists within the lookback window). `pre` is nil when the
/// user skipped the pre-session prompt.
struct SessionPainPair {
    let pre: PainMeasurement?
    let post: PainMeasurement
}

/// Numeric view of pain change. `delta = post - pre` (negative means
/// improvement). `delta` is nil when no pre log exists.
struct SessionPainDelta: Equatable {
    let pre: Int?
    let post: Int
    let delta: Int?
}

/// Walks the log list and matches each POST to the most recent PRE
/// that occurred before it within `lookback`. Defaults to one hour,
/// matching Android.
func pairSessionPainLogs(
    _ logs: [PainMeasurement],
    lookback: TimeInterval = 60 * 60
) -> [SessionPainPair] {
    let pres = logs
        .filter { $0.context == PainContext.preSession }
        .sorted { $0.recordedAt < $1.recordedAt }
    let posts = logs
        .filter { $0.context == PainContext.postSession }
        .sorted { $0.recordedAt > $1.recordedAt }

    return posts.map { post in
        let pre = pres.last { pre in
            pre.recordedAt <= post.recordedAt &&
                post.recordedAt.timeIntervalSince(pre.recordedAt) <= lookback
        }
        return SessionPainPair(pre: pre, post: post)
    }
}

func sessionPainDelta(_ pair: SessionPainPair) -> SessionPainDelta {
    let pre = pair.pre?.painLevel
    let post = pair.post.painLevel
    return SessionPainDelta(
        pre: pre,
        post: post,
        delta: pre.map { post - $0 }
    )
}
